app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt

```diff
@@ -324,6 +330,59 @@ class ExpressionControlFlowTest {
         )
     }
 
+    @Test
+    fun `globals - read and write`() {
+        val context = ExpressionContext(
+            setOf("x"),
+            globals = setOf("x")
+        )
+
+        val read = context.createCfg("x" asVarExprIn context)
+        assertTrue(
+            read hasSameStructureAs
+                IntermediateFormTreeNode.MemoryRead(
+                    IntermediateFormTreeNode.Add(
+                        IntermediateFormTreeNode.MemoryLabel("globals"),
+                        IntermediateFormTreeNode.Const(0)
+                    )
+                ).toCfg()
+        )
+
+        val write = context.createCfg(Expression.NumberLiteral(10), "x" asVarIn context)
+        assertTrue(
+            write hasSameStructureAs
+                IntermediateFormTreeNode.MemoryWrite(
+                    IntermediateFormTreeNode.Add(
+                        IntermediateFormTreeNode.MemoryLabel("globals"),
+                        IntermediateFormTreeNode.Const(0)
+                    ),
+                    IntermediateFormTreeNode.Const(10)
+                ).toCfg()
+        )
+    }
+
+    @Test
+    fun `globals - multiple variables`() {
+        val context = ExpressionContext(
+            setOf("x", "y", "z"),
+            globals = setOf("x", "y", "z")
+        )
+
+        val cfgs = context.varNames.map { context.createCfg(it asVarExprIn context) }
+
+        fun cfgForOffset(offset: Long) =
+            IntermediateFormTreeNode.MemoryRead(
+                IntermediateFormTreeNode.Add(
+                    IntermediateFormTreeNode.MemoryLabel("globals"),
+                    IntermediateFormTreeNode.Const(offset)
+                )
+            ).toCfg()
+
+        for (i in 0 until 3) {
+            assertTrue(cfgs.filter { it hasSameStructureAs cfgForOffset(i * 4.toLong()) }.size == 1)
+        }
+    }
+
```

Comment:
I don't think we should test details of `GlobalVariablesAccessGenerator` (such as exact name of the label or offset values) in this file. I would suggest making separate test file for this class (which would be relatively short I guess) and in this file just ensure that the proper object is used for global variables (maybe by adding parameter `isGlobal` to `DummyRead/Write`...). That corresponds to my comment about ControlFlow.kt - passing `GlobalVariablesAccessGenerator` as an argument should make testing easier.

Reply:
> passing `GlobalVariablesAccessGenerator` as an argument

I wrote it in a little misleading way - of course I meant passing `VariableAccessGenerator` interface.